src/org/intellij/erlang/compilation/ErlangPrepareDependenciesCompileTask.java

```diff
@@ -128,161 +128,118 @@ private static ErlangModuleBuildOrderDescriptor getModuleBuildOrderInner(Module
+        return new ErlangFileDescriptor(filePath, semiGraph.getDependencies(filePath));
+      }
+    });
+  }
 
-    private ErlangModulesSorter(Collection<ErlangFile> modules, List<String> globalParseTransforms) {
-      myModules = modules;
-      myGlobalParseTransforms = globalParseTransforms;
-    }
 
-    public static List<ErlangFileDescriptor> sort(Collection<ErlangFile> erlangModules,
-                                                  List<String> globalParseTransforms) throws CyclicDependencyFoundException {
-      ErlangModulesSorter sorter = new ErlangModulesSorter(erlangModules, globalParseTransforms);
-      return sorter.getSortedModules();
+  private static class ErlangFilesDependencyGraph implements GraphGenerator.SemiGraph<String> {
+    private final HashMap<String, String> myNamesToPathsMap = ContainerUtil.newHashMap();
+    private final HashMap<String, Collection<String>> myPathsToDependenciesMap = ContainerUtil.newHashMap();
+
+    public ErlangFilesDependencyGraph(@NotNull Collection<ErlangFile> erlangModules,
+                                      @NotNull Collection<ErlangFile> erlangHeaders,
+                                      @NotNull List<String> globalParseTransforms) {
+      buildNamesMap(erlangModules);
+      buildForHeaders(erlangHeaders);
+      buildForNodes(erlangModules, globalParseTransforms);
     }
 
-    private GraphGenerator<Node> createModulesGraph() {
-      return GraphGenerator.create(new ErlangModulesDependencyGraph(myModules, myGlobalParseTransforms));
+    private void buildForNodes(@NotNull Collection<ErlangFile> erlangModules,
+                               @NotNull List<String> globalParseTransforms) {
+      List<String> globalPtNodes = getModulePaths(globalParseTransforms);
+      for (ErlangFile erlangModule : erlangModules) {
+        Set<String> dependencies = ContainerUtil.newHashSet();
+        dependencies.addAll(getPathsFromNames(ErlangPsiImplUtil.getAppliedParseTransformModuleNames(erlangModule)));
+        dependencies.addAll(getPathsFromNames(ErlangPsiImplUtil.getImplementedBehaviourModuleNames(erlangModule)));
+        dependencies.addAll(ErlangPsiImplUtil.getIncludedHeadersFilePaths(erlangModule));
+        dependencies.addAll(getPathsFromNames(globalPtNodes));
+        myPathsToDependenciesMap.put(getErlangFilePath(erlangModule), dependencies);
+      }
     }
 
-    private List<ErlangFileDescriptor> getSortedModules() throws CyclicDependencyFoundException {
-      GraphGenerator<Node> graph = createModulesGraph();
-      DFSTBuilder<Node> builder = new DFSTBuilder<Node>(graph);
-      builder.buildDFST();
-      if (!builder.isAcyclic()) {
-        throw new CyclicDependencyFoundException();
+    private void buildForHeaders(@NotNull Collection<ErlangFile> erlangHeaders) {
+      for (ErlangFile header : erlangHeaders) {
+        myPathsToDependenciesMap.put(getErlangFilePath(header), ContainerUtil.newArrayList(ErlangPsiImplUtil.getIncludedHeadersFilePaths(header)));
```

Comment:
You have inconsistencies when you build this graph.

For module files, you compute include, parse_transform, and behaviour dependencies recursively from all included files. But for included files you only consider inclusion dependencies. I'm not sure if this results in incorrect behaviour - please see if it does. Anyway, there is a performance hit here: Include dependencies computation is quadratic.